### ISSUE TITLE
[Enhancement req.] Toggle for gamelist text length to extend to right beyond 50% of screen

### ISSUE THREAD

**Issue by john-iv** [Mataniko/IV-Play#4]
F1 UI toggle to allow game list text to flow to the right over the 50% screen barrier (if history.dat is not being displayed).  Looks better when UI is 800px wide and using small snaps (this allows text to overflow art elements on the right)  This was a feature during the initial building of the app.

**Comment by retroarchgirl** [comment]
I don't know if I understand current code enough to implement this, but I can have a look.

Can you make a screenshot by using a photo editor on a screenshot of the current version to indicate clearly:

- how far the text will go (compared to current version)
- the overflow of text over art elements because I don't see the point here. You want WHITE text over a snapshot? Otherwise it's not visible with other colors and even white can give problems on certain snapshots. 

It's not clear for me how it should look according to you.
I see @Mataniko never replied, so maybe this is not possible or very hard in current code?

Thanks

**Comment by john-iv** [comment]
There is a hard text cut-off at 50% horizontal of the size of the window.  It appears that it's not artview constrained as you can see with this screenshot using 'Super Large' art view, the text flows over it.  My initial request was to allow the cut off text to extend, like in Real Bout Fatal Fury 2 to extend the rest of the way.  So based on the size of the window, if it were small, it could get pretty ugly I suppose.  But it might be interesting to mock-up and see it.  It could also potentially be blocked by the clickable art field itself (to toggle art views, snaps, history, mameinfo etc.) if text extends into it.

![1](https://user-images.githubusercontent.com/9202235/236078466-55b626f1-7a61-4b02-9948-96d560abd92d.png)

**Comment by retroarchgirl** [comment]
So basically you want an F1 option to indicate in % how far the strings go?
So total control over it, no matter how big the window is?

I still don't understand why you would be interested in snapshots with text on top?

- This means 100% is the whole window, from left to right even if you made it very small?


- You want to enlarge the window in width (so snapshot is further away from the text) and then be able to expand the 50% limit to a value you choose?
Is that the reason?

**Comment by john-iv** [comment]
The use case is merely to allow for the lengthiest gamename + manufacturer + year to be displayed without cutoff without necessarily widening the window (which is really the easiest answer I suppose).  Normally one would think that people wouldn't want a super small window where this would look poor (see attached).

Is there a place in the code that specifies that the text is to stop at 50% of the window?  I'll have to see what game/machine in the UI is the longest.  A 60% setting may handle it.

**Comment by retroarchgirl** [comment]
I have to look at this.

I think (without now looking at code) that the developer just used 1/2 of the window.

It would be either

- use  1/2 of the window
- use a % you choose

So it can't be both

Or you want a check 'if game name string is X chars long" switch to the % choice?

I have to check later. I also find the code is not super clean and sometimes it's a mess to change things.
I pass if it's the case here

**Comment by retroarchgirl** [comment]
I took a SUPER fast look and then it's for another time. It's late here

`private string ValidateRowLength(string text, int offset)
    {
      if (string.IsNullOrEmpty(text))
        return "";

      Graphics g = CreateGraphics();
      float textWidth = g.MeasureString(text, Font).Width;
      bool trimmed = false;
      while (textWidth + offset + ArtOffset > ((float)ClientRectangle.Width / 2))
      {
        trimmed = true;
        text = text.Remove(text.Length - 1, 1);
        textWidth = g.MeasureString(text, Font).Width;
      }

      if (trimmed && text.Length > 3)
      {
        text = text.Remove(text.Length - 3, 3);
        text = text + "...";
      }
      return text;
    }`

So it's 1/2 of the window as I said

`textWidth + offset + ArtOffset > ((float)ClientRectangle.Width / 2)`

**Comment by retroarchgirl** [comment]
I think this would be not good to implement because

- checking each string for length => slowing UI down
- what do you get as result for the little amount of games that are too long? (compared to the speed decrease you get?)
- this code ValidateRowLength is also run in the highlight drawing so you check the string length there then TOO (so twice)

So I suggest making a fixed % for ALL games as a F1 option

**Comment by john-iv** [comment]
This is another instance where I think putting this in the iv-play.cfg is appropriate as a percentage value; these are arcane tweaks that don't need to clutter the F1 UI.  I wonder though if setting it to 75% would interfere with the drawing of the history.xml and mameinfo.dat area.  Might have text over text which wouldn't look too good.

Looks like the tvgames/spg2xx_jakks_sharp.cpp games have the longest name combos.

This is low priority.

**Comment by retroarchgirl** [comment]
Yes it will mess text up.You can't do nothing zbout it since you don't know where the other history text is (it's based on the sentences in the text)
That's why the developer made it 'safe' at 50% not to enter the 'danger zone'...

**Comment by john-iv** [comment]
Is the history / mameinfo area bounded, can it be shrunk or is it also a percentage of window width?
It's the lowest priority, don't expend any energy on it right now.

**Comment by retroarchgirl** [comment]
The rectangle in which the text is displayed is defined by

`RectangleF rectangleF = new RectangleF(
        ((float)ClientRectangle.Width / 2),
        ArtOffset,
        ((float)ClientRectangle.Width / 2) - ArtOffset,
        sizeRect.Height
      );`

It starts at point ( (float)ClientRectangle.Width / 2 , ArtOffset ) and the width is (float)ClientRectangle.Width / 2) - ArtOffset , while the height is sizeRect.Height

Don't ask me where the numbers of sizeRect come from (the "- 11"), defined as

`var stringFormat = new StringFormat();
var sizeRect = g.MeasureString(
        _currentInfoText,
        Settings.Default.info_font,
        new SizeF(((float)ClientRectangle.Width / 2) - 11, 0),
        stringFormat
      );`

To me it's clear the UI things are hardcoded everywhere. Only the original developer will now specific things. He tried and hard coded things to text don't overlap. No other explanation

The "stringFormat" containing the text is drawn in a rectangle area "rectangleF"  using the font and color you choose in F1

So it also uses the 50% border to calculate. You can shrink it but to what values? It doesn't know the length of the game names.
So the 50% border is put in there to keep things safe. Lazy coding :)

I don't know a dynamic way to let the text area shrink when game names are longer, and expand when they are shorter.
Certainly not in the current code!

**Comment by john-iv** [comment]
Closing, implementation and UI design considerations.
